### ISSUE TITLE
ML Model Benchmarking: 48% → 90% prediction accuracy through systematic model comparison

### ISSUE THREAD

**Issue by holden** [razrfly/cinegraph#659]
## Objective

Systematically benchmark multiple ML models to find the optimal approach for predicting 1001 Movies list membership. Current accuracy is stuck at **48%** across all 5 hand-tuned weight profiles (44.5%–48.8% spread). The goal is to reach **90%** through data-driven model selection rather than manual weight tuning.

This is a dedicated breakout of **Phase 5B** from #658.

## Why Hand-Tuning Failed

The 5 named weight profiles span only **4.3 percentage points**. This means the problem isn't _which weights_ we pick — it's that linear weighted scoring over 6 criteria with massive data gaps (festival data: 4.5% coverage in 1920s, Metacritic nonexistent pre-1980) cannot separate list movies from non-list movies. We need models that can:

1. Handle missing data natively (not penalize movies for lacking festival/critic data)
2. Discover non-linear feature interactions (e.g., "high IMDb votes + on Sight & Sound list" is a stronger signal than either alone)
3. Weight features differently by decade (festival data is noise pre-1950, strong signal post-1990)

## Feature Matrix

Export from PostgreSQL to Parquet. Each row = one movie.

### Existing Features (from current 6 criteria)
| Feature | Source | Coverage | Notes |
|---------|--------|----------|-------|
| `mob_score` | IMDb/TMDb/RT audience | ~95% | Strongest universal signal |
| `ivory_tower_score` | Metacritic/RT tomatometer | ~40% | Missing pre-1980 |
| `festival_recognition_score` | festival_nominations | ~65% | 4.5% in 1920s, 95.6% in 2010s |
| `cultural_impact_score` | ROI, votes, genre | ~70% | Budget/revenue often 0 for classics |
| `technical_innovation_score` | Technical Oscar/fest noms | ~50% | Category name matching is fragile |
| `auteur_recognition_score` | Director's list count | ~98% | Most reliable criterion |

### New Features to Add
| Feature | Source | Coverage | Hypothesis |
|---------|--------|----------|-----------|
| `canonical_overlap_count` | `canonical_sources` JSONB | **100%** | Movie on S&S + TSPDT + Criterion = near-certain 1001 member. 9 lists in DB. |
| `canonical_overlap_weighted` | Weighted by list prestige | **100%** | S&S Top 250 > TSPDT Top 1000 > others |
| `imdb_votes_log` | `external_metrics` | ~95% | Log-scaled cultural staying power proxy |
| `imdb_rating` | `external_metrics` | ~95% | Raw 0-10 rating, universal |
| `decade` | `release_date` | **100%** | Era-aware modeling |
| `has_festival_data` | Boolean flag | **100%** | Diagnostic: does missingness predict list membership? |
| `has_critic_data` | Boolean flag | **100%** | Same diagnostic |
| `director_film_count` | `movie_credits` | ~98% | Prolific directors vs one-hit wonders |
| `years_since_release` | Derived | **100%** | Older confirmed films = higher cultural proof |

**Label:** `is_on_1001_list` — binary, from `canonical_sources ? '1001_movies'`

### 🥇 Experiment 1: XGBoost (Try First)

**Why first:** Proven in volfefe for tabular data with mixed feature types. Handles missing values natively (`use_missing=True`). Fast iteration (~5-30s per run on M4 Mac Studio). Feature importance via SHAP is interpretable.

**Configuration:**
```python
XGBClassifier(
    n_estimators=1000,
    max_depth=6,
    learning_rate=0.01,
    early_stopping_rounds=50,
    eval_metric='aucpr',  # better for imbalanced classes
    scale_pos_weight=ratio,  # ~1001 list movies are minority class
    tree_method='hist',  # fast on CPU
)
```

**Validation:** Stratified K-Fold CV (k=10) + per-decade holdout (train on all decades except one, predict the held-out decade).

**Success criteria:**
- ✅ **>70% accuracy** = promising, worth optimizing
- ✅ **>80% accuracy** = strong model, move to Phase C
- ✅ **>90% accuracy** = target achieved
- ❌ **<55% accuracy** = features insufficient, need more data enrichment
- ❌ **SHAP shows `canonical_overlap_count` dominates all other features** = circular (lists predict lists), need to test without it

**Key analysis:** Run SHAP to identify top-3 features. If `canonical_overlap` is #1, re-run without it to see floor accuracy from other signals alone.

**Timeline:** 1 day setup + export, 1 day experiments

---

### 🥈 Experiment 2: Isolation Forest (Anomaly Framing)

**Why second:** Reframes the problem — instead of "predict list membership," ask "which movies are anomalously high-quality?" Unsupervised, so no risk of overfitting to list labels. Proven AUC=0.77 in volfefe for detecting outliers in tabular data.

**Configuration:**
```python
IsolationForest(
    n_estimators=500,
    contamination=0.05,  # ~5% of movies are "must-see"
    max_samples='auto',
    random_state=42,
)
```

**Validation:** Rank all movies by anomaly score, check what % of top-N are actual 1001 list members (Precision@K).

**Success criteria:**
- ✅ **Precision@1001 > 60%** = anomaly framing captures quality signal
- ✅ **Precision@1001 > 80%** = strong unsupervised baseline
- ❌ **Precision@1001 < 40%** = anomaly framing doesn't fit this domain (quality ≠ anomaly)

**Why it might fail:** 1001 Movies list is curated by committee, not purely quality-driven. Anomaly detection finds statistical outliers, but "important film" might not be statistically anomalous.

**Timeline:** 0.5 day (can reuse XGBoost feature export)

---

### 🥉 Experiment 3: Logistic Regression (Interpretable Baseline)

**Why third:** Dead simple, fully interpretable, and gives us a _floor_ for what linear models can do. If LogReg beats our hand-tuned 48%, it proves the problem was never about model complexity — just about picking the right features. Also gives coefficients that translate directly to Elixir weights.

**Configuration:**
```python
LogisticRegression(
    C=1.0,
    class_weight='balanced',
    max_iter=1000,
    solver='lbfgs',
)
```

**Validation:** Same stratified K-Fold + per-decade holdout as XGBoost.

**Success criteria:**
- ✅ **>60% accuracy** = proves feature selection matters more than model complexity; coefficients become new Elixir weights directly
- ✅ **>75% accuracy** = linear model sufficient, no need for complex ML in production
- ❌ **<50% accuracy** = confirms non-linear interactions are essential (justifies XGBoost)

**Why it might fail:** If the signal is in feature _interactions_ (e.g., "high votes AND on canonical lists"), linear models can't capture this without manual feature engineering.

**Timeline:** 0.5 day

---

### Experiment 4: Random Forest

**Why:** Ensemble of decision trees, good for understanding feature importance without SHAP overhead. Natural comparison point for XGBoost (if RF ≈ XGBoost, the simpler model wins).

**Configuration:**
```python
RandomForestClassifier(
    n_estimators=500,
    max_depth=None,  # let trees grow deep
    class_weight='balanced_subsample',
    min_samples_leaf=5,
    n_jobs=-1,
)
```

**Success criteria:**
- ✅ **Within 3% of XGBoost** = use RF for simplicity
- ❌ **>5% below XGBoost** = gradient boosting needed, stick with XGBoost

**Timeline:** 0.5 day

---

### Experiment 5: Autoencoder (Reconstruction Error)

**Why:** Learns a compressed representation of "what a typical movie looks like." Movies that reconstruct poorly = unusual = potentially list-worthy. Different signal than supervised models.

**Architecture (from volfefe pattern):**
```python
# Input → 32 → 16 → 8 (bottleneck) → 16 → 32 → Output
# Loss: MSE reconstruction error
# Epochs: 200, batch_size=4096, lr=1e-3, early_stopping patience=20
```

**Success criteria:**
- ✅ **Top-1001 by reconstruction error captures >50% list movies** = useful ensemble component
- ❌ **<35% capture** = reconstruction error doesn't correlate with "importance"

**Why it might fail:** volfefe showed autoencoder signal degrades with dataset growth. With ~100K movies, the bottleneck may not separate quality from randomness.

**Timeline:** 1 day (PyTorch setup + training)

---

### Experiment 6: Ensemble (Best-Of Combination)

**Why:** Combine top 2-3 models' predictions. In volfefe, ensemble averaging (IF + AE) achieved best unsupervised AUC=0.80, beating any individual model.

**Method:** Simple average or stacked logistic regression over individual model scores.

**Success criteria:**
- ✅ **Ensemble > best individual model by >3%** = worth the complexity
- ❌ **Ensemble ≈ best individual** = just use the single best model

**Timeline:** 0.5 day (after Experiments 1-5 complete)

---

## Experiment Tracking

Mirror volfefe's MLflow setup:

```
scripts/
  ml/
    export_features.py          # PostgreSQL → Parquet
    exp01_xgboost.py            # XGBoost + SHAP
    exp02_isolation_forest.py   # Anomaly framing
    exp03_logistic_regression.py # Linear baseline
    exp04_random_forest.py      # Ensemble trees
    exp05_autoencoder.py        # Reconstruction error
    exp06_ensemble.py           # Best-of combination
    shared/
      data_loader.py            # Feature loading + imputation
      evaluator.py              # Precision@K, accuracy, per-decade metrics
    requirements.txt
```

Each experiment logs to MLflow:
- Parameters (hyperparameters, feature set version)
- Metrics (overall accuracy, per-decade accuracy, AUC-ROC, Precision@K)
- Artifacts (SHAP plots, confusion matrices, score distributions)

### Primary Metric: Historical Validation Accuracy
Same methodology as current Elixir backtest — for each decade, predict top-N movies (where N = actual 1001 list count), compare against ground truth.

### Secondary Metrics
- **Per-decade accuracy** — Are we strong everywhere or only post-1990?
- **AUC-ROC** — Ranking quality (does the model rank list movies higher than non-list?)
- **Precision@K** — Of top-K predictions, how many are actually on the list?
- **SHAP feature importance** — Which features drive predictions? (guards against circular logic)

### Validation Strategy
1. **Stratified 10-Fold CV** — Standard, comparable across models
2. **Decade-Holdout CV** — Train on 9 decades, predict 1 (tests temporal generalization)
3. **Leave-One-List-Out** — Remove `canonical_overlap` feature to test if signal is circular

### Accuracy Tiers

| Accuracy | Verdict | Action |
|----------|---------|--------|
| <50% | ❌ Failure | Features insufficient. Go back to data enrichment. |
| 50–65% | ⚠️ Marginal | Some signal found. Check if new features (canonical_overlap) help. |
| 65–80% | ✅ Good | Usable model. Identify which features to add to Elixir scoring. |
| 80–90% | ✅✅ Strong | Excellent. Translate weights back to production criteria. |
| >90% | ✅✅✅ Target | Ship it. Build decade-aware weight profiles from model coefficients. |

## Estimated Timeline

| Day | Work |
|-----|------|
| 1 | Feature export script (PostgreSQL → Parquet), data_loader, evaluator |
| 2 | Exp 1: XGBoost + SHAP analysis |
| 3 | Exp 2: Isolation Forest + Exp 3: Logistic Regression |
| 4 | Exp 4: Random Forest + Exp 5: Autoencoder |
| 5 | Exp 6: Ensemble + final comparison + write-up |

**Total: ~5 days** on Mac Studio M4 (64GB RAM). XGBoost on 100K movies is trivially fast on CPU.

## Dependencies
- Phase 5A lens validation (#658) should ideally complete first to ensure score quality
- PostgreSQL access from Python (connectorx + polars)
- Python 3.11+ with: `polars`, `connectorx`, `xgboost`, `scikit-learn`, `shap`, `mlflow`, `torch`, `matplotlib`

## Definition of Done
- [ ] All 6 experiments run with results logged to MLflow
- [ ] Comparison table with accuracy per model per decade
- [ ] SHAP analysis identifying top features (with and without canonical_overlap)
- [ ] Recommendation: which model to use and which features to add to Elixir scoring
- [ ] If >80% achieved: draft PR updating `criteria_scoring.ex` with ML-discovered weights

## References
- Parent issue: #658
- volfefe experiment structure: `experiments/01_isolation_forest/` through `experiments/16_*/` — 19 experiments with MLflow tracking
- volfefe shared evaluation: `experiments/shared/evaluator.py` (Precision@K, AUC-ROC, permutation tests)

**Comment by holden** [comment]
## Experiment 01 — XGBoost Results

### Setup
- **Dataset**: 1,003,273 movies, 1,256 positive labels (on 1001 list) — severe class imbalance (~1:797)
- **V1 features**: 6 lens scores only (mob, ivory_tower, festival_recognition, cultural_impact, technical_innovation, auteur)
- **V2 features**: V1 + canonical_overlap_count, imdb_rating, imdb_votes_log, decade, has_festival_data, has_critic_data, director_film_count, years_since_release

### Results

| Model | 10-fold CV AUC | AUC (20% holdout) | P@500 (holdout) | P@1001 (holdout) |
|---|---|---|---|---|
| Baseline (hand-tuned) | — | — | — | 0.48 |
| XGBoost V1 (lens only) | **0.9996 ± 0.0003** | 1.0000 | 0.502 | 0.251 |
| XGBoost V2 (all features) | **1.0000 ± 0.0000** | 1.0000 | 0.502 | 0.251 |

### Interpreting the Numbers

**AUC=1.0 is the headline finding.** On the 20% holdout, both models ranked every single one of the ~251 held-out 1001-list movies above all ~200K non-list movies. Perfect discrimination.

The low holdout P@K is **not** a weakness — it's a denominator problem. With only ~251 positives in a 200K holdout, Precision@1001 can't exceed ~0.25. For an apples-to-apples comparison with the 48% baseline (which is evaluated against all 1,256 positives), we'd need cross-validated predictions on the full dataset. That's a quick follow-up.

### Data Leakage Warning ⚠️

V2's `canonical_overlap_count` feature (number of other canonical lists a movie appears on) is essentially a near-perfect proxy for the label — the same curated films dominate all major lists. It inflates metrics and should be treated as a separate experiment or dropped from the core feature set. V1 (lens scores only) is the cleaner signal.

V1 AUC=0.9996 in 10-fold CV without any list-membership features is the more interesting result. It means the six lens dimensions alone almost perfectly discriminate 1001-list films from the rest — though this raises its own question: **are the lens scores themselves calibrated against curated lists?** Worth auditing.

### Next Steps

1. **Compute cross-validated P@1001** on the full dataset for a fair comparison to the 48% baseline
2. **Audit feature leakage**: check whether lens scores are correlated with the 1001-list label by design
3. **Run Experiment 02** (logistic regression) to see whether XGBoost's complexity is necessary or if the signal is essentially linear
4. If AUC=1.0 holds up after leakage audit, the bottleneck shifts from model accuracy to **lens score coverage** (many movies have NULL scores)

### Environment
```
XGBoost 2.1.4, scikit-learn 1.6.1, Python 3.9, cinegraph/scripts/ml/.venv
```

**Comment by holden** [comment]
SUPERSEDED by #682 — post-leakage-audit successor with experiment roadmap.
